### PR TITLE
Skip recorded symlinks in --setperms (RhBug:1900662)

### PR DESCRIPTION
**Pull request by dmnks**
Since we record the symlink permission bits (777) in the database and
chmod(1) affects the pointed-to file (not the symlink itself), applying
these bits to anything on the file system would either be pointless (a
symlink) or an error (a ghost file present as a regular file but
recorded as a symlink in the database).

The former is already handled with the -L test, this commit handles the
latter.

### CODE REVIEW

**Comment by dmnks** [comment]
Note that we might as well consider skipping ghost files altogether; the permissions of a ghost file recorded at build time may not be relevant to the actual file system state when the package is installed, but I'm not sure if that's a valid assumption.

**Review by Conan-Kudo — APPROVED** [review]

**Comment by dmnks** [comment]
I guess I should also include a test case (I already have one that I used locally to test this), so hold on one sec.

**Comment by dmnks** [comment]
Dragons from the past could be awaken if we don't do this right, see e.g:
38c2f6e160d5ed3e9c3a266139c7eb2632724c15
0d83637769b8a122b1e80f2e960ea1bbae8b4f10
(Thanks @pmatilai for ^^^)

**Comment by Conan-Kudo** [comment]
We have DRAGONS now as a label? 😆

**Comment by dmnks** [comment]
Yeah, it's always about ghosts and dragons, anyway... :smile:

**Comment by dmnks** [comment]
I was wondering if we're not addressing this issue in the wrong place; as per POSIX, symlink permission bits are unspecified, yet we record those in the rpmdb. Shouldn't we instead leave them unspecified (0) in the rpmdb as well? That would prevent this kind of confusion with `--setperms` which then tries to applies these symlink "permissions" to `%ghost` files that happen to be regular files.

**Comment by pmatilai** [comment]
> Explicitly skip recorded symlinks in --setperms

This is the right thing to do, regardless of anything else. Permission bits are not applicable for symlinks so we simply should *never* try to set them, no matter what.

It's the ghosts and goblins part where things get confusing, if you let the mind wander there :ghost:

**Comment by pmatilai** [comment]
So... I think the change as such is fine, but as this actually isn't ghost-specific, I'd reword the commit summar somewhat. Something like "Skip recorded symlinks in --setperms (RhBug:1900662)", the message itself elaborates on the various cases, including %ghost.

**Comment by dmnks** [comment]
> This is the right thing to do, regardless of anything else. Permission bits are not applicable for symlinks so we simply should _never_ try to set them, no matter what.

True, it really is a no-brainer. What's key here is the file *type* (a binary value in our case - whether it was a symlink or not) for which these permissions were recorded, and we know it (a little awkwardly through `%FILELINKTO`), so it should be taken into account in `--setperms`, indeed.

> It's the ghosts and goblins part where things get confusing, if you let the mind wander there ghost

As for the most recent dragon encounter (the two commits mentioned above), the first commit 38c2f6e was about avoiding a `chmod` on files/directories that don't exist. The regression it caused (fixed in the second commit 0d83637) was due to the wrong conditional logic (missing parenthesis). Other than that, there doesn't seem to be anything there that interferes with the case we're fixing here, so in that regard, we should be safe.

**Comment by dmnks** [comment]
> So... I think the change as such is fine, but as this actually isn't ghost-specific, I'd reword the commit summar somewhat. Something like "Skip recorded symlinks in --setperms (RhBug:1900662)", the message itself elaborates on the various cases, including %ghost.

Yep, I'll fix that (also the message itself needs a bit of care).

**Comment by dmnks** [comment]
OK, so I've come to the realization that what we're actually dealing with here is *changing file types* where symlinks are just a special case of that (and one that's probably the most prominent). We want to skip any such files (whose type has changed), though. I've pushed a generalized commit, please let me know your thoughts.

**Comment by pmatilai** [comment]
Now you're really inviting them dragons... trying to be overly clever in the cramped alias space is exactly what got into trouble the last time.

I certainly preferred the simpler version which only dealt with symlinks: we should skip recorded links, regardless of whether they've changed or not. Leave the fancy business to the eventual C implementation :)

**Comment by dmnks** [comment]
> Now you're really inviting them dragons... trying to be overly clever in the cramped alias space is exactly what got us into trouble the last time.

Valid point :)

> I certainly preferred the simpler version which only dealt with symlinks: we should skip recorded links, regardless of whether they've changed or not. Leave the fancy business to the eventual C implementation :)

Cool, no problem! I'll push the previous commit (with the updated msg) in a moment.

**Comment by dmnks** [comment]
Commit pushed, how about this one?

**Comment by pmatilai** [comment]
Works for me!
